compose_rl/data/dataloader.py

```diff
@@ -117,9 +125,36 @@ def build_preference_dataloader(
     finegrained_preference_dataset_collate_fn,
 )
 
+
+def get_num_tokens_in_batch_online(
+    batch: dict[str, Any],
+) -> int:
+    """Get the number of tokens in batch including prompt + generated tokens.
+
+    Uses action_mask and prompt_len for precise counting when available.
+    """
+    if 'action_mask' in batch and 'prompt_len' in batch:
+        prompt_len_tokens = batch['prompt_len'].sum().item()
+        generated_items = batch['action_mask'].sum().item()
+        padding_tokens = batch['action_mask'].numel() - generated_items
+        log.info(f'prompt tokens in batch: {prompt_len_tokens}')
+        log.info(f'generated tokens in batch: {generated_items}')
+        log.info(
+            f'unused generation (padding) tokens in batch: {padding_tokens}',
+        )
+        return int(prompt_len_tokens + generated_items)
+    else:
+        log.warning(
+            'No action_mask/prompt_len in batch. ' +
+            'Using default value of 0 for num_tokens_in_batch.',
+        )
+        return 0
```

Comment:
Using 0 since [composer default is also 0](https://sourcegraph.prod.databricks-corp.com/mosaicml/composer/-/blob/composer/core/data_spec.py?L306) and we don't want to raise an error here so I logged a warning instead.